Cal sorting

hsergi opened the pull request:
## Describe your changes
I have implemented an algorithm that sorts out frames collected for either k-gain, non-linearity or em-gain calibration. The case of em-gain calibration is partly implemented: there is a mock set generated, though it is not yet sorted by ``sorting.py``. This is an agreed work to do, and it is not necessary for v1.1

The sorting algorithm adds a keyword OBSTYPE to the sorted files with values MNFRAME, KGAIN, NONLIN, according to the calibration type selected (Notice that the frames to generate a mean frames are always sorted for any calibration type)

## Type of change

I added two new functions:

``sorting.py``: sorting algorithm with I/O Datasets and the calibration type to be considered
The algorithm respects the constraints of the calibration data as defined in https://collaboration.ipac.caltech.edu/pages/viewpage.action?spaceKey=romancoronagraph&title=Calibration+functions+description


## How to run the code
``python test_sorting.py``

## Issues

In a mac system, set at least ulimit -n 800 before running the code

## Checklist before requesting a review
- [ x] I have linted my code: 9.02/10
- [ x] I have verified that all unit tests pass in a clean environment and added new unit tests, as needed
- [ x] I have verified that all docstrings are properly formatted and added new documentation, as needed

Commit:
Test script for sorting calibration data types for EM-gain, k-gain and non-linearity

Start

Commit:
Adding NL and EM gain constraints

Commit:
test Dataset for cal sorting of mean frame, k-gain, non-linearity and em-gain (vs DAC) ready

Commit:
I/O are list of filenames

Commit:
Mean frame sorting and templates

Commit:
Name and I/O changes

As required in https://github.com/roman-corgi/corgidrp/issues/231

Commit:
Kgain first algo

Commit:
kgain algo

semaphoreP committed:
update VISTYPE for FLT, ASTROM, DARK

semaphoreP committed:
fix some bugs to get unit tests passing

semaphoreP committed:
add option (not used) to process multiple recipes with the same input dataset

semaphoreP committed:
add ability to pass in arbitrary filepaths for recipe templates

Commit:
Algo for mean frame, kgain and nonlin implemented (TODO: em-gain)

Commit:
Doc string spaces

Commit:
More doc strings

semaphoreP commented on corgidrp/sorting.py:
```diff
@@ -0,0 +1,232 @@
+
+    return filename[idx_0:idx_1]
+
+def sorting(
```
this name should be a little less generic (e.g., `sort_pupilimg_frames()`). In the future, we may have multipel sorting functions here for different data types.

hsergi replied:
I asked you and I used you suggested name: ``sorting.py``. Mine was ``cal_soring``. I found it too generic too. I'll change to this one.

semaphoreP commented on corgidrp/sorting.py:
```diff
@@ -0,0 +1,232 @@
+        
+    # EM-gain: TODO
+    elif cal_type.lower() == 'em-gain' or cal_type.lower() == 'emgain':
+        print('Considering low EM-gain: TODO')
```
shouldn't this be raising an exception?

hsergi replied:
As Marie has mentioned (slack 11/6, 2:08pm): "after chatting with Vanessa, it sounds like we will not need to implement EM-gain sorting after all, which is good news!" And "All that matters is that we correctly select frames from nonlin and kgain, while excluding the frames only useful for EM-gain."

The test function does generate cal frames for EM-gain (vs DAC) and sorts k-gain and non-linearity sets from them. It does not attempt to create a dataset for EM-gain until it becomes a requirement. Let's stick to clear tasks compatible with current specifications in a complex, whose specifications are not very much fixed yet (OBSTYPE example, filename convention, DATETIME, etc) rather than approaching it as a flexible, personal project

semaphoreP commented on corgidrp/sorting.py:
```diff
@@ -0,0 +1,232 @@
+        raise Exception('Unrecognized calibration type (expected k-gain, non-lin or em-gain)')
+
+    # Return Datafrane with mean frame and cal type
+    return data.Dataset(mean_frame_list + cal_frame_list)
```
before returnign the dataset, it will be good to add a HISTORY entry to the ext_hdr (like other step functions) that describe what sorting was done.

hsergi replied:
@semaphoreP  I do not know enough about corgidrp. Not sure how to do this. My experience is with step functions that are a derivative of Image and have in its creation the pri_hdr and ext_hdr keywords. I have reviewed the step functions in corgidrp and that's what I've found.

In this case, I'm using Dataset.  Also, each frame of this dataset has a modified primary header with a specific value of OBSTYPE, and EXPTIME, CMDGAIN in its extension header. I am not sure if creating the final Dataset object based on a single ext header is going to change the values of all the frames.

Given the time left to finish several other tasks related with this PR, can you write how it should be?

semaphoreP replied:
if you have a dataset object, just run `dataset.update_after_processing_step("what was done in this step")`. Doing this will write "what was done in this step" into the history of every image in the dataset.

hsergi replied:
Oh yeah. Long time I had not seen it.

hsergi replied:
Done

semaphoreP commented on tests/test_sorting.py:
```diff
@@ -0,0 +1,367 @@
+    dataset_cal = data.Dataset(filename_list)
+
+    # Apply sorting algorithm and check results (maybe output of sorting is
+    # mean frame and the type used as input? Instead of them all. Then, check
```
the comment says we should check the properties, but there are no checks in this unit test. The unit test should have some asserts at the end that checks the outputs are what are expected.

hsergi replied:
I submitted the PR as soon as I could for your review. Will do.

hsergi replied:
Added a full set of tests consistent with the aim this PR for DRP v1.1

semaphoreP commented on tests/test_sorting.py:
```diff
@@ -0,0 +1,367 @@
+    hdul.writeto(filename, overwrite = True)
+    return filename
+
+# Main code
```
A unit test needs to be inside a function starting with "test_". It can't just be in a script like this.

hsergi replied:
Done

semaphoreP commented on tests/test_sorting.py:
```diff
@@ -0,0 +1,367 @@
+    return cmdgain_list, exptime_list
+
+def get_cmdgain_exptime_kgain(
+    exptime_sec=[0.077, 0.770, 1.538, 2.308, 3.077, 3.846, 4.615, 5.385, 6.154,
```
this is minor, but python keywords shouldn't host non-primitive objects like lists. Instead the python best practice is to set the keyword to None e.g., `exptime_sec=None`, and then at the start of the function, have a block of code that is like `if exptime_sec is None: exptime_sec = [...]`.

hsergi replied:
I set None and did not set the lists inside the functions because they must be defined for the assertions later on and it forces the user to know what's been set as a representative set of calibration data for K-gain, non-linearity and EM-gain vs DAC. 

The latter is important when adding more possible combinations in a post DRP-v1.1 review of the sorting algorithm of datasets than the one in the current test, which is based on TVAC and a draft version of the detailed implementation of the corresponding CAR (Commissioning Activity Report). It may change in the future and/or there are other distinct combinations to test

semaphoreP commented on corgidrp/sorting.py:
```diff
@@ -0,0 +1,232 @@
+def extract_frame_id(filename):
+    """
+    Extract frame ID from an L1 filename. Structure is assumed to be ending
+    like '..._frame_id.fits' where frame_id is a series of digits
```
is it confirmed that the last element of a filename will be a frame ID?

hsergi replied:
The plan is to be compatible with IIT. Now, it is. This is a comment for the next version

hsergi replied:
I'll resolve this once I have the time to post an updated list of to-do's in https://github.com/roman-corgi/corgidrp/issues/226

semaphoreP commented on corgidrp/sorting.py:
```diff
@@ -0,0 +1,232 @@
+                cal_frame_list += [frame]
+                n_kgain += 1
+
+        print(f'K-gain has {n_kgain} unity frames with exposure times', exptime_cons[idx_kgain_0:idx_kgain_1+1], f'seconds with each {count_cons[idx_kgain_0]} frames each')
```
this print statement and the nonlin one below seem like good things to write into the header

hsergi replied:
I added it to the HISTORY.

Example when run for K-gain:
Dataset to calibrate K-GAIN. A sorting algorithm based on the constraints that NFRAMES, EXPTIME and CMDGAIN have when collecting calibration data for K-gain, Non-linearity and EM-gain vs DAC was applied to an input dataset from PUPILIMG visit files.The result is that K-gain has 110 unity frames with exposure times [0.077, 0.77, 1.538, 2.308, 3.077, 3.846, 4.615, 5.385, 6.154, 6.923, 7.692, 8.462, 9.231, 10.0, 10.769, 11.538, 12.308, 13.077, 13.846, 14.615, 15.385, 1.538] seconds with 5 frames each.

Example when run for non-linearity:
Dataset to calibrate NON-LIN. A sorting algorithm based on the constraints that NFRAMES, EXPTIME and CMDGAIN have when collecting calibration data for K-gain, Non-linearity and EM-gain vs DAC was applied to an input dataset from PUPILIMG visit files.The result is that Non-linearity has 242 frames with gains [1.65, 5.24, 8.6, 16.7, 27.5, 45.26, 87.5, 144.1, 237.26, 458.7, 584.4].

semaphoreP replied:
Perfect. This is exactly what we are looking to put into the history.

semaphoreP reviewed (commented):
What is the plan to add this to the e2e test + recipes? Do you have time this week do work on that as well?

semaphoreP commented:
Please also fix the docstrings to make the tests pass.

semaphoreP committed:
queue nonlin and kgain together in pupilimg recipe

semaphoreP committed:
add walker edge case for no VISTYPE for IIT data

hsergi commented:
I'll try again the docstrings issue. It's quite annoying tbh because I mimicked other modules I wrote that passed and still does not. Do you have a reference for us to follow if we can't have a minor flexibity?

semaphoreP commented:
> I'll try again the docstrings issue. It's quite annoying tbh because I mimicked other modules I wrote that passed and still does not. Do you have a reference for us to follow if we can't have a minor flexibity?

I think this one is pretty self-explanatory if you read the error message produced by the build:

![image](https://github.com/user-attachments/assets/f4fbb2b9-0e0f-42a9-9648-d16e8cadd15e)

Basically your variable name in your function declaration `em_emgain` doesn't match the variable name you put in the docstring `em_gain`. You can also run these checks locally on your computer before you push the code (see the README file for the repo).

Commit:
Merge branch 'obstype-switch' into cal_sorting

hsergi commented:
Issue when running e2e tests after merging branch obstype-switch: https://github.com/roman-corgi/corgidrp/issues/250

Commit:
Function names changes plus on-going tests

Commit:
None in test function. Test function with expected syntax

Commit:
Improved K-gain test function

Commit:
Minor doc string typo fixed

Commit:
Removing unnecessary dataset copies

Commit:
Trying to comply with pytest

Commit:
Testing build

Commit:
Update test_sort_pupilimg_frames.py

Commit:
Unit tests completed

semaphoreP committed:
fix nonlin e2e test

Commit:
History string

Commit:
Updated sorted dataset with HISTORY

Commit:
Merge remote-tracking branch 'origin/obstype-switch' into cal_sorting

Commit:
Cleaning up before dealing with the remaining issues of the e2e test for NL

semaphoreP commented on tests/test_sort_pupilimg_frames.py:
```diff
@@ -0,0 +1,569 @@
+# values of exposure times
+
+# Values for mean frame
+cmdgain_mean_frame, exptime_mean_frame = get_cmdgain_exptime_mean_frame(
```
In general, if you have this much setup code, it's good practice to put it in a setup function. pytest has a `setup_module()`  function that you can define. this runs setup code for a .py file. I would put all this code in there, and use the python `global` declaration inside that function to identify the variables you want to grab for the unit tests. 

```
def setup_module():
    global dataset_w_change, dataset_wo_change # this will make these variables you create in this function accessible to other functions
    # your code here
```

hsergi replied:
Indeed I tried myself, because that amount of block code is quite ugly as-is, though I did not know the precise syntax of the function

hsergi replied:
Done

Commit:
As is now (two issues)

Non-linearity:

nonlin_dataset_2_stack() in calibrate_nonlin.py: Data are there and split. However, the substacks have some minor issue that makes it fail in l.252 (if np.sum(len_list) != len(cal_arr)).

K-gain:
It seems the algo has a mismatch of one subset of unity gain frames

semaphoreP reviewed (commented):
just one request to put your data generation code from your unit tests into a specifically named function called `setup_module()`. This is better than leaving all the data generation code at the base of the module.

Commit:
Kgain: crashes for some I/O FITS. Unrelated to this PR?

Commit:
First try/pass in kgain_e2e with the cal library is confusing

Commit:
setup_module() for test_sorting_frames()

Commit:
Typo with substack

Commit:
Update calibrate_nonlin.py

hsergi commented:
@kjl0025 in case he has some ideas:

I've implemented all the changes for the sorting algorithm and have written a unit test.

As for the e2e, it's been long and confusing. I realized that if I remove the cosmic ray detection from the recipe on non linearity, and I leave the other modules, the walker ends up successfully though the nonlinearity table is not good with errors of a few percent, instead of single precision. Not sure where the error comes from. The sorting algorithm seems to find the right combo: Dataset to calibrate NON-LIN. A sorting algorithm based on the constraints that NFRAMES, EXPTIME and CMDGAIN have when collecting calibration data for K-gain, Non-linearity and EM-gain vs DAC was applied to an input dataset from PUPILIMG visit files.The result is that Non-linearity has 242 frames with gains [1.649999976158142, 5.239999771118164, 8.600000381469727, 16.70000076293945, 27.5, 45.2599983215332, 87.5, 144.1000061035156, 237.2599945068359, 458.7000122070312, 584.4000244140625]
Consistent with https://collaboration.ipac.caltech.edu/pages/viewpage.action?spaceKey=romancoronagraph&title=Calibration+functions+description

If I add the cosmic ray, I get this long error message that I am not sure where it is coming from (https://github.com/roman-corgi/corgidrp/issues/251)

Running walker
Traceback (most recent call last):
  File "nonlin_e2e.py", line 206, in <module>
    test_nonlin_cal_e2e(args.tvacdata_dir, args.output_dir)
  File "nonlin_e2e.py", line 139, in test_nonlin_cal_e2e
    walker.walk_corgidrp(nonlin_l1_list, '', e2eoutput_path, "l1_to_l2a_nonlin.json")
  File "/Users/srhildeb/Library/CloudStorage/OneDrive-JPL/Documents/GitHub/corgidrp/corgidrp/walker.py", line 77, in walk_corgidrp
    recipes = autogen_recipe(filelist, outputdir, template=template)
  File "/Users/srhildeb/Library/CloudStorage/OneDrive-JPL/Documents/GitHub/corgidrp/corgidrp/walker.py", line 153, in autogen_recipe
    _fill_in_calib_files(step, this_caldb, first_frame)
  File "/Users/srhildeb/Library/CloudStorage/OneDrive-JPL/Documents/GitHub/corgidrp/corgidrp/walker.py", line 194, in _fill_in_calib_files
    best_cal_file = this_caldb.get_calib(ref_frame, calib_dtype)
  File "/Users/srhildeb/Library/CloudStorage/OneDrive-JPL/Documents/GitHub/corgidrp/corgidrp/caldb.py", line 314, in get_calib
    return dtype(calib_filepath)
  File "/Users/srhildeb/Library/CloudStorage/OneDrive-JPL/Documents/GitHub/corgidrp/corgidrp/data.py", line 853, in __init__
    super().__init__(data_or_filepath, err=err, pri_hdr=pri_hdr, ext_hdr=ext_hdr, err_hdr=err_hdr)
  File "/Users/srhildeb/Library/CloudStorage/OneDrive-JPL/Documents/GitHub/corgidrp/corgidrp/data.py", line 271, in __init__
    with fits.open(data_or_filepath, ignore_missing_simple=True) as hdulist:
  File "/Users/srhildeb/miniconda3/envs/cgi_etc/lib/python3.7/site-packages/astropy/io/fits/hdu/hdulist.py", line 174, in fitsopen
    lazy_load_hdus, ignore_missing_simple, **kwargs)
  File "/Users/srhildeb/miniconda3/envs/cgi_etc/lib/python3.7/site-packages/astropy/io/fits/hdu/hdulist.py", line 411, in fromfile
    lazy_load_hdus=lazy_load_hdus, **kwargs)
  File "/Users/srhildeb/miniconda3/envs/cgi_etc/lib/python3.7/site-packages/astropy/io/fits/hdu/hdulist.py", line 1064, in _readfrom
    fileobj = _File(fileobj, mode=mode, memmap=memmap, cache=cache)
  File "/Users/srhildeb/miniconda3/envs/cgi_etc/lib/python3.7/site-packages/astropy/utils/decorators.py", line 536, in wrapper
    return function(*args, **kwargs)
  File "/Users/srhildeb/miniconda3/envs/cgi_etc/lib/python3.7/site-packages/astropy/io/fits/file.py", line 170, in __init__
    self._open_filename(fileobj, mode, overwrite)
  File "/Users/srhildeb/miniconda3/envs/cgi_etc/lib/python3.7/site-packages/astropy/io/fits/file.py", line 558, in _open_filename
    self._file = fileobj_open(self.name, IO_FITS_MODES[mode])
  File "/Users/srhildeb/miniconda3/envs/cgi_etc/lib/python3.7/site-packages/astropy/io/fits/util.py", line 390, in fileobj_open
    return open(filename, mode, buffering=0)
FileNotFoundError: [Errno 2] No such file or directory: '/Users/srhildeb/Library/CloudStorage/OneDrive-JPL/Documents/GitHub/corgidrp/tests/e2e_tests/l1_to_kgain_output/CGI_EXCAM_L1_0000051867_kgain.fits'

hsergi commented:
And this is for K-gain: Considering K-gain:
Dataset to calibrate K-GAIN. A sorting algorithm based on the constraints that NFRAMES, EXPTIME and CMDGAIN have when collecting calibration data for K-gain, Non-linearity and EM-gain vs DAC was applied to an input dataset from PUPILIMG visit files.The result is that K-gain has 110 unity frames with exposure times [0.077, 0.769, 1.538, 2.308, 3.077, 3.846, 4.615, 5.385, 6.154, 6.923, 7.692, 8.462, 9.231, 10.0, 10.769, 11.538, 12.308, 13.077, 13.846, 14.615, 15.385, 1.538] seconds with 5 frames each.

It crashes :

Running walker
Traceback (most recent call last):
  File "kgain_e2e.py", line 153, in <module>
    test_l1_to_kgain(tvacdata_dir, outputdir)
  File "kgain_e2e.py", line 111, in test_l1_to_kgain
    walker.walk_corgidrp(ordered_filelist, "", kgain_outputdir, template="l1_to_kgain.json")
  File "/Users/srhildeb/Library/CloudStorage/OneDrive-JPL/Documents/GitHub/corgidrp/corgidrp/walker.py", line 77, in walk_corgidrp
    recipes = autogen_recipe(filelist, outputdir, template=template)
  File "/Users/srhildeb/Library/CloudStorage/OneDrive-JPL/Documents/GitHub/corgidrp/corgidrp/walker.py", line 153, in autogen_recipe
    _fill_in_calib_files(step, this_caldb, first_frame)
  File "/Users/srhildeb/Library/CloudStorage/OneDrive-JPL/Documents/GitHub/corgidrp/corgidrp/walker.py", line 194, in _fill_in_calib_files
    best_cal_file = this_caldb.get_calib(ref_frame, calib_dtype)
  File "/Users/srhildeb/Library/CloudStorage/OneDrive-JPL/Documents/GitHub/corgidrp/corgidrp/caldb.py", line 314, in get_calib
    return dtype(calib_filepath)
  File "/Users/srhildeb/Library/CloudStorage/OneDrive-JPL/Documents/GitHub/corgidrp/corgidrp/data.py", line 853, in __init__
    super().__init__(data_or_filepath, err=err, pri_hdr=pri_hdr, ext_hdr=ext_hdr, err_hdr=err_hdr)
  File "/Users/srhildeb/Library/CloudStorage/OneDrive-JPL/Documents/GitHub/corgidrp/corgidrp/data.py", line 271, in __init__
    with fits.open(data_or_filepath, ignore_missing_simple=True) as hdulist:
  File "/Users/srhildeb/miniconda3/envs/cgi_etc/lib/python3.7/site-packages/astropy/io/fits/hdu/hdulist.py", line 174, in fitsopen
    lazy_load_hdus, ignore_missing_simple, **kwargs)
  File "/Users/srhildeb/miniconda3/envs/cgi_etc/lib/python3.7/site-packages/astropy/io/fits/hdu/hdulist.py", line 411, in fromfile
    lazy_load_hdus=lazy_load_hdus, **kwargs)
  File "/Users/srhildeb/miniconda3/envs/cgi_etc/lib/python3.7/site-packages/astropy/io/fits/hdu/hdulist.py", line 1064, in _readfrom
    fileobj = _File(fileobj, mode=mode, memmap=memmap, cache=cache)
  File "/Users/srhildeb/miniconda3/envs/cgi_etc/lib/python3.7/site-packages/astropy/utils/decorators.py", line 536, in wrapper
    return function(*args, **kwargs)
  File "/Users/srhildeb/miniconda3/envs/cgi_etc/lib/python3.7/site-packages/astropy/io/fits/file.py", line 170, in __init__
    self._open_filename(fileobj, mode, overwrite)
  File "/Users/srhildeb/miniconda3/envs/cgi_etc/lib/python3.7/site-packages/astropy/io/fits/file.py", line 558, in _open_filename
    self._file = fileobj_open(self.name, IO_FITS_MODES[mode])
  File "/Users/srhildeb/miniconda3/envs/cgi_etc/lib/python3.7/site-packages/astropy/io/fits/util.py", line 390, in fileobj_open
    return open(filename, mode, buffering=0)
FileNotFoundError: [Errno 2] No such file or directory: '/Users/srhildeb/Library/CloudStorage/OneDrive-JPL/Documents/GitHub/corgidrp/tests/e2e_tests/l1_to_kgain_output/CGI_EXCAM_L1_0000051867_kgain.fits'

Commit:
Last tests (Tables are ~% different) worse than single precision

Commit:
e2e NL pass (though w/o CR. See commit message)

Without, they pass well.

With:

FileNotFoundError: [Errno 2] No such file or directory: '/Users/srhildeb/Library/CloudStorage/OneDrive-JPL/Documents/GitHub/corgidrp/tests/e2e_tests/l1_to_kgain_output/CGI_EXCAM_L1_0000051867_kgain.fits'

Commit:
e2e: NL passes. Kgain maybe? Issues with kgain_e2e unrelated to this pR

e2e NL: here is an issue is the CR stepi when using "KGain" : "AUTOMATIC, OPTIONAL"

e2e Kgain: in addition to the previous issues, when removing it kgain_e2e crashes: FileNotFoundError: [Errno 2] No such file or directory: 'l1_to_kgain_output/CGI_EXCAM_L1_0000051731_kgain.fits'

Commit:
Typo in an exception (flake)

Commit:
Update nonlin_e2e.py

Commit:
Update nonlin_e2e.py

Commit:
Update nonlin_e2e.py

Commit:
One more blank line

Commit:
Fix Kgain/nonlin unit test pass

Commit:
Moving sorting to the top. e2e NL passes. corgidrp_caldb.csv in .corgidrp can be tricky

corgidrp_caldb.csv can trigger crashes for some (tbd) reasons. I did not change or delete manually. In case of issues, it seems safe to delete all entries or start anew

Commit:
Minor changes

semaphoreP committed:
fix path issues in e2e test

semaphoreP committed:
nonlin e2e saves to own dir and fix filepath issues
